data-structures/cosmos/basic_types.md

```diff
@@ -0,0 +1,90 @@
+# Tendermint and Cosmos Data Structure Basic Types
+
+```ipldsch
+# Uint is a non-negative integer
+type Uint bytes
+
+# The main purpose of HexBytes is to enable HEX-encoding for json/encoding.
+type HexBytes bytes
+
+# Address is a type alias of a slice of bytes
+# An address is calculated by hashing the public key using sha256
+# and truncating it to only use the first 20 bytes of the slice
+type Address HexBytes
+
+# Hash is a type alias of a slice of 32 bytes
+type Hash HexBytes
+
+# Time represents a unix timestamp with nanosecond granularity
+type Time struct {
+    Seconds Uint
+    Nanoseconds Uint
+}
+
+# Version captures the consensus rules for processing a block in the blockchain,
+# including all blockchain data structures and the rules of the application's
+# state transition machine.
+type Version struct {
+	Block Uint
+	App   Uint
+}
+
+# BlockIDFlag is a single byte flag
+type BlockIDFlag enum {
+  | BlockIDFlagUnknown ("0")
+  | BlockIDFlagAbsent ("1")
+  | BlockIDFlagCommit ("2")
+  | BlockIDFlagNil ("3")
+} representation int
+```
+
+## Params
```

Comment:
I'm new to this repo, how would upgrades be handled? The tendermint repo has a couple changes in mind to core data structures

Reply:
Updates to the data structures would need to be reflected in the schemas here and the implementations of the schema types and their marshallers and unmarshallers (in this repo for the Go implementation: https://github.com/vulcanize/go-codec-dagcosmos).